File: apps/server/routers/integrations.py
"""
Integrations router for external service APIs (VOICEVOX, etc.)
"""

from fastapi import APIRouter, HTTPException, Query, UploadFile, File
from fastapi.responses import FileResponse
from typing import List, Optional
from pathlib import Path
import httpx
import logging
import os
import shutil
import uuid

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/integrations", tags=["integrations"])

# Directory for uploaded models (relative to server)
# In production, this should be configurable
UPLOAD_DIR = Path(__file__).parent.parent.parent.parent / "apps" / "web" / "public" / "models"
UPLOAD_DIR.mkdir(parents=True, exist_ok=True)

# Directory for uploaded animations
ANIMATIONS_DIR = Path(__file__).parent.parent.parent.parent / "apps" / "web" / "public" / "animations"
ANIMATIONS_DIR.mkdir(parents=True, exist_ok=True)

# Directory for generated audio files
AUDIO_DIR = Path(__file__).parent.parent.parent.parent / "apps" / "server" / "audio_output"
AUDIO_DIR.mkdir(parents=True, exist_ok=True)

# Allowed file extensions
ALLOWED_EXTENSIONS = {".vrm", ".png", ".jpg", ".jpeg", ".gif", ".webp"}
ALLOWED_ANIMATION_EXTENSIONS = {".fbx", ".glb", ".gltf"}


@router.get("/voicevox/speakers")
async def get_voicevox_speakers(
    host: str = Query(default="http://localhost:50021", description="VOICEVOX host URL")
):
    """
    Fetch available speakers from VOICEVOX engine.
    Returns a flattened list of speakers with their styles.
    """
    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(f"{host.rstrip('/')}/speakers")
            response.raise_for_status()
            speakers_data = response.json()

            # Flatten speakers and styles into a simple list
            speakers = []
            for speaker in speakers_data:
                speaker_name = speaker.get("name", "Unknown")
                for style in speaker.get("styles", []):
                    speakers.append({
                        "id": style.get("id"),
                        "name": speaker_name,
                        "style": style.get("name", "Normal"),
                        "label": f"{speaker_name} ({style.get('name', 'Normal')})",
                    })

            return {"speakers": speakers}

    except httpx.ConnectError:
        raise HTTPException(
            status_code=503,
            detail=f"Cannot connect to VOICEVOX at {host}. Make sure VOICEVOX is running."
        )
    except httpx.HTTPStatusError as e:
        raise HTTPException(
            status_code=e.response.status_code,
            detail=f"VOICEVOX API error: {e.response.text}"
        )
    except Exception as e:
        logger.error(f"Error fetching VOICEVOX speakers: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch speakers: {str(e)}"
        )


@router.get("/voicevox/health")
async def check_voicevox_health(
    host: str = Query(default="http://localhost:50021", description="VOICEVOX host URL")
):
    """Check if VOICEVOX engine is running and accessible."""
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            response = await client.get(f"{host.rstrip('/')}/version")
            response.raise_for_status()
            version = response.text.strip('"')
            return {"status": "healthy", "version": version, "host": host}
    except Exception as e:
        return {"status": "unhealthy", "error": str(e), "host": host}


@router.get("/audio/{filename}")
async def serve_audio(filename: str):
    """
    Serve generated audio files.
    Audio files are stored in the server audio_output directory.
    """
    # Security: allow only .wav files with no path traversal
    if ".." in filename or "/" in filename or "\\" in filename:
        raise HTTPException(status_code=400, detail="Invalid filename")
    if not filename.lower().endswith(".wav"):
        raise HTTPException(status_code=400, detail="Invalid filename")

    audio_path = (AUDIO_DIR / filename).resolve()
    if not audio_path.exists() or audio_path.parent != AUDIO_DIR.resolve():
        raise HTTPException(status_code=404, detail="Audio file not found")

    return FileResponse(
        path=str(audio_path),
        media_type="audio/wav",
        filename=filename
    )


@router.post("/models/upload")
async def upload_model(file: UploadFile = File(...)):
    """
    Upload a VRM model or image file.
    Returns the URL path to access the uploaded file.
    """
    if not file.filename:
        raise HTTPException(status_code=400, detail="No filename provided")

    # Check file extension
    ext = Path(file.filename).suffix.lower()
    if ext not in ALLOWED_EXTENSIONS:
        raise HTTPException(
            status_code=400,
            detail=f"File type not allowed. Allowed types: {', '.join(ALLOWED_EXTENSIONS)}"
        )

    # Generate unique filename to avoid conflicts
    unique_id = str(uuid.uuid4())[:8]
    safe_filename = f"{unique_id}_{Path(file.filename).stem}{ext}"
    file_path = UPLOAD_DIR / safe_filename

    try:
        # Save file
        with open(file_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)

        # Return the URL path (served by backend API)
        url_path = f"/api/integrations/models/file/{safe_filename}"

        logger.info(f"Uploaded model: {safe_filename}")

        return {
            "success": True,
            "filename": safe_filename,
            "url": url_path,
            "size": file_path.stat().st_size
        }

    except Exception as e:
        logger.error(f"Error uploading file: {e}")
        # Clean up on error
        if file_path.exists():
            file_path.unlink()
        raise HTTPException(status_code=500, detail=f"Failed to upload file: {str(e)}")


@router.get("/models")
async def list_models():
    """
    List all uploaded models.
    """
    try:
        models = []
        for file_path in UPLOAD_DIR.iterdir():
            if file_path.is_file() and file_path.suffix.lower() in ALLOWED_EXTENSIONS:
                models.append({
                    "filename": file_path.name,
                    "url": f"/api/integrations/models/file/{file_path.name}",
                    "size": file_path.stat().st_size,
                    "type": "vrm" if file_path.suffix.lower() == ".vrm" else "image"
                })

        return {"models": models}

    except Exception as e:
        logger.error(f"Error listing models: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to list models: {str(e)}")


@router.delete("/models/{filename}")
async def delete_model(filename: str):
    """
    Delete an uploaded model.
    """
    # Security: validate filename
    if ".." in filename or "/" in filename or "\\" in filename:
        raise HTTPException(status_code=400, detail="Invalid filename")

    file_path = UPLOAD_DIR / filename

    if not file_path.exists():
        raise HTTPException(status_code=404, detail="File not found")

    try:
        file_path.unlink()
        logger.info(f"Deleted model: {filename}")
        return {"success": True, "message": f"Deleted {filename}"}

    except Exception as e:
        logger.error(f"Error deleting file: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to delete file: {str(e)}")


@router.get("/models/file/{filename}")
async def serve_model(filename: str):
    """
    Serve uploaded model files directly from backend.
    This allows access without Next.js dev server restart.
    """
    # Security: validate filename
    if ".." in filename or "/" in filename or "\\" in filename:
        raise HTTPException(status_code=400, detail="Invalid filename")

    file_path = UPLOAD_DIR / filename

    if not file_path.exists():
        raise HTTPException(status_code=404, detail="File not found")

    # Determine media type
    ext = file_path.suffix.lower()
    media_types = {
        ".vrm": "model/gltf-binary",
        ".png": "image/png",
        ".jpg": "image/jpeg",
        ".jpeg": "image/jpeg",
        ".gif": "image/gif",
        ".webp": "image/webp",
    }
    media_type = media_types.get(ext, "application/octet-stream")

    return FileResponse(
        path=str(file_path),
        media_type=media_type,
        filename=filename
    )


# Animation file endpoints
@router.post("/animations/upload")
async def upload_animation(file: UploadFile = File(...)):
    """
    Upload an animation file (FBX from Mixamo, etc.).
    Returns the URL path to access the uploaded file.
    """
    if not file.filename:
        raise HTTPException(status_code=400, detail="No filename provided")

    # Check file extension
    ext = Path(file.filename).suffix.lower()
    if ext not in ALLOWED_ANIMATION_EXTENSIONS:
        raise HTTPException(
            status_code=400,
            detail=f"File type not allowed. Allowed types: {', '.join(ALLOWED_ANIMATION_EXTENSIONS)}"
        )

    # Generate unique filename to avoid conflicts
    unique_id = str(uuid.uuid4())[:8]
    safe_filename = f"{unique_id}_{Path(file.filename).stem}{ext}"
    file_path = ANIMATIONS_DIR / safe_filename

    try:
        # Save file
        with open(file_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)

        # Return the URL path
        url_path = f"/api/integrations/animations/file/{safe_filename}"

        logger.info(f"Uploaded animation: {safe_filename}")

        return {
            "success": True,
            "filename": safe_filename,
            "url": url_path,
            "size": file_path.stat().st_size
        }

    except Exception as e:
        logger.error(f"Error uploading animation: {e}")
        if file_path.exists():
            file_path.unlink()
        raise HTTPException(status_code=500, detail=f"Failed to upload file: {str(e)}")


@router.get("/animations")
async def list_animations():
    """
    List all uploaded animations.
    """
    try:
        animations = []
        for file_path in ANIMATIONS_DIR.iterdir():
            if file_path.is_file() and file_path.suffix.lower() in ALLOWED_ANIMATION_EXTENSIONS:
                animations.append({
                    "filename": file_path.name,
                    "url": f"/api/integrations/animations/file/{file_path.name}",
                    "size": file_path.stat().st_size,
                    "type": file_path.suffix.lower()[1:]  # fbx, glb, etc.
                })

        return {"animations": animations}

    except Exception as e:
        logger.error(f"Error listing animations: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to list animations: {str(e)}")


@router.delete("/animations/{filename}")
async def delete_animation(filename: str):
    """
    Delete an uploaded animation.
    """
    if ".." in filename or "/" in filename or "\\" in filename:
        raise HTTPException(status_code=400, detail="Invalid filename")

    file_path = ANIMATIONS_DIR / filename

    if not file_path.exists():
        raise HTTPException(status_code=404, detail="File not found")

    try:
        file_path.unlink()
        logger.info(f"Deleted animation: {filename}")
        return {"success": True, "message": f"Deleted {filename}"}

    except Exception as e:
        logger.error(f"Error deleting animation: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to delete file: {str(e)}")


@router.get("/animations/file/{filename}")
async def serve_animation(filename: str):
    """
    Serve uploaded animation files.
    """
    if ".." in filename or "/" in filename or "\\" in filename:
        raise HTTPException(status_code=400, detail="Invalid filename")

    file_path = ANIMATIONS_DIR / filename

    if not file_path.exists():
        raise HTTPException(status_code=404, detail="File not found")

    ext = file_path.suffix.lower()
    media_types = {
        ".fbx": "application/octet-stream",
        ".glb": "model/gltf-binary",
        ".gltf": "model/gltf+json",
    }
    media_type = media_types.get(ext, "application/octet-stream")

    return FileResponse(
        path=str(file_path),
        media_type=media_type,
        filename=filename
    )


# VRM Model file endpoints
ALLOWED_MODEL_EXTENSIONS = {".vrm"}


@router.post("/models/upload")
async def upload_model(file: UploadFile = File(...)):
    """
    Upload a VRM model file.
    Returns the URL path to access the uploaded file.
    """
    if not file.filename:
        raise HTTPException(status_code=400, detail="No filename provided")

    # Check file extension
    ext = Path(file.filename).suffix.lower()
    if ext not in ALLOWED_MODEL_EXTENSIONS:
        raise HTTPException(
            status_code=400,
            detail=f"File type not allowed. Allowed types: {', '.join(ALLOWED_MODEL_EXTENSIONS)}"
        )

    # Generate unique filename to avoid conflicts
    unique_id = str(uuid.uuid4())[:8]
    safe_filename = f"{unique_id}_{Path(file.filename).stem}{ext}"
    file_path = UPLOAD_DIR / safe_filename

    try:
        # Save file
        with open(file_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)

        # Return the URL path (relative to public folder)
        url_path = f"/models/{safe_filename}"

        logger.info(f"Uploaded model: {safe_filename}")

        return {
            "success": True,
            "filename": safe_filename,
            "url": url_path,
            "size": file_path.stat().st_size
        }

    except Exception as e:
        logger.error(f"Error uploading model: {e}")
        if file_path.exists():
            file_path.unlink()
        raise HTTPException(status_code=500, detail=f"Failed to upload file: {str(e)}")


@router.get("/models")
async def list_models():
    """
    List all uploaded VRM models.
    """
    try:
        models = []
        for file_path in UPLOAD_DIR.iterdir():
            if file_path.is_file() and file_path.suffix.lower() in ALLOWED_MODEL_EXTENSIONS:
                models.append({
                    "filename": file_path.name,
                    "url": f"/models/{file_path.name}",
                    "size": file_path.stat().st_size,
                })

        return {"models": models}

    except Exception as e:
        logger.error(f"Error listing models: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to list models: {str(e)}")


@router.delete("/models/{filename}")
async def delete_model(filename: str):
    """
    Delete an uploaded VRM model.
    """
    if ".." in filename or "/" in filename or "\\" in filename:
        raise HTTPException(status_code=400, detail="Invalid filename")

    file_path = UPLOAD_DIR / filename

    if not file_path.exists():
        raise HTTPException(status_code=404, detail="File not found")

    try:
        file_path.unlink()
        logger.info(f"Deleted model: {filename}")
        return {"success": True, "message": f"Deleted {filename}"}

    except Exception as e:
        logger.error(f"Error deleting model: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to delete file: {str(e)}")

File: apps/server-ts/src/routes/integrations.ts
/**
 * Integrations API Routes - VOICEVOX, file management.
 *
 * Ported from Python apps/server/routers/integrations.py
 */

import { mkdir, readdir, unlink } from "node:fs/promises";
import { basename, extname, join } from "node:path";
import { Hono } from "hono";
import type { ContentfulStatusCode } from "hono/utils/http-status";
import { getProjectRoot } from "../engine/plugin-loader";

const app = new Hono();

const PROJECT_ROOT = getProjectRoot();
const UPLOAD_DIR = process.env.UPLOAD_DIR || join(PROJECT_ROOT, "apps", "web", "public", "models");
const ANIMATIONS_DIR =
  process.env.ANIMATIONS_DIR || join(PROJECT_ROOT, "apps", "web", "public", "animations");
const AUDIO_DIR = process.env.AUDIO_DIR || join(PROJECT_ROOT, "apps", "server-ts", "audio_output");
const DEFAULT_VOICEVOX_HOST = "http://localhost:50021";
const MAX_MODEL_UPLOAD_BYTES = 100 * 1024 * 1024;
const MAX_ANIMATION_UPLOAD_BYTES = 100 * 1024 * 1024;
const ALLOWED_VOICEVOX_HOSTS = new Set(["localhost", "127.0.0.1", "0.0.0.0", "::1", "[::1]"]);

const ALLOWED_EXTENSIONS = new Set([".vrm", ".png", ".jpg", ".jpeg", ".gif", ".webp"]);
const ALLOWED_ANIMATION_EXTENSIONS = new Set([".fbx", ".glb", ".gltf"]);

interface VoicevoxStyle {
  id: number;
  name?: string;
}

interface VoicevoxSpeaker {
  name?: string;
  styles?: VoicevoxStyle[];
}

function validateVoicevoxHost(rawHost?: string): {
  host: string;
  error?: string;
} {
  const value = rawHost ?? DEFAULT_VOICEVOX_HOST;

  let parsed: URL;
  try {
    parsed = new URL(value);
  } catch {
    return { host: DEFAULT_VOICEVOX_HOST, error: "Invalid host URL" };
  }

  if (!["http:", "https:"].includes(parsed.protocol)) {
    return {
      host: DEFAULT_VOICEVOX_HOST,
      error: "Only http/https hosts are allowed",
    };
  }

  if (!ALLOWED_VOICEVOX_HOSTS.has(parsed.hostname)) {
    return {
      host: DEFAULT_VOICEVOX_HOST,
      error: "Only localhost VOICEVOX hosts are allowed",
    };
  }

  return { host: `${parsed.protocol}//${parsed.host}` };
}

function sanitizeUploadStem(filename: string): string {
  const base = basename(filename);
  const stem = base.replace(/\.[^.]+$/, "");
  const sanitized = stem
    .replace(/[^A-Za-z0-9_-]/g, "_")
    .replace(/_+/g, "_")
    .replace(/^_+|_+$/g, "");
  return sanitized || "upload";
}

function sanitizeHeaderFilename(filename: string): string {
  return filename.replace(/[^A-Za-z0-9._-]/g, "_");
}

function isVoicevoxUnavailableError(err: unknown): boolean {
  if (!(err instanceof Error)) return false;

  if (err.name === "AbortError" || err.name === "TimeoutError") {
    return true;
  }

  const withCause = err as Error & { cause?: { code?: string } };
  return err.name === "TypeError" || withCause.cause?.code === "ECONNREFUSED";
}

// Ensure directories exist
await mkdir(UPLOAD_DIR, { recursive: true }).catch(() => {});
await mkdir(ANIMATIONS_DIR, { recursive: true }).catch(() => {});
await mkdir(AUDIO_DIR, { recursive: true }).catch(() => {});

// ─── VOICEVOX ─────────────────────────────

app.get("/voicevox/speakers", async (c) => {
  const { host, error } = validateVoicevoxHost(c.req.query("host"));
  if (error) {
    return c.json({ detail: error }, 400);
  }

  try {
    const response = await fetch(`${host}/speakers`, {
      signal: AbortSignal.timeout(10000),
    });
    if (!response.ok) {
      return c.json(
        { detail: `VOICEVOX API error: ${response.statusText}` },
        response.status as ContentfulStatusCode,
      );
    }

    const speakersData = (await response.json()) as VoicevoxSpeaker[];
    const speakers = speakersData.flatMap((speaker) =>
      (speaker.styles ?? []).map((style) => ({
        id: style.id,
        name: speaker.name ?? "Unknown",
        style: style.name ?? "Normal",
        label: `${speaker.name ?? "Unknown"} (${style.name ?? "Normal"})`,
      })),
    );

    return c.json({ speakers });
  } catch (err) {
    if (isVoicevoxUnavailableError(err)) {
      return c.json(
        {
          detail: `Cannot connect to VOICEVOX at ${host}. Make sure VOICEVOX is running.`,
        },
        503,
      );
    }
    console.error("Failed to fetch VOICEVOX speakers:", err);
    return c.json({ detail: "Failed to fetch speakers" }, 500);
  }
});

app.get("/voicevox/health", async (c) => {
  const { host, error } = validateVoicevoxHost(c.req.query("host"));
  if (error) {
    return c.json({ detail: error }, 400);
  }

  try {
    const response = await fetch(`${host}/version`, {
      signal: AbortSignal.timeout(5000),
    });
    if (!response.ok) throw new Error(response.statusText);
    const version = (await response.text()).replace(/"/g, "");
    return c.json({ status: "healthy", version, host });
  } catch (err) {
    console.error("VOICEVOX health check failed:", err);
    return c.json({ status: "unhealthy", error: "VOICEVOX unavailable", host }, 503);
  }
});

// ─── Audio Serving ────────────────────────

app.get("/audio/:filename", async (c) => {
  const filename = c.req.param("filename");

  // Security validation
  if (filename.includes("..") || filename.includes("/") || filename.includes("\\")) {
    return c.json({ detail: "Invalid filename" }, 400);
  }
  if (!filename.toLowerCase().endsWith(".wav")) {
    return c.json({ detail: "Invalid filename" }, 400);
  }

  const filePath = join(AUDIO_DIR, filename);
  const file = Bun.file(filePath);
  if (!(await file.exists())) {
    return c.json({ detail: "Audio file not found" }, 404);
  }

  const safeHeaderFilename = sanitizeHeaderFilename(filename);

  return new Response(file, {
    headers: {
      "Content-Type": "audio/wav",
      "Content-Disposition": `inline; filename="${safeHeaderFilename}"`,
    },
  });
});

// ─── Model Upload/List/Delete/Serve ───────

app.post("/models/upload", async (c) => {
  const formData = await c.req.formData();
  const file = formData.get("file") as File | null;
  if (!file || !file.name) {
    return c.json({ detail: "No file provided" }, 400);
  }

  const ext = extname(file.name).toLowerCase();
  if (!ALLOWED_EXTENSIONS.has(ext)) {
    return c.json(
      {
        detail: `File type not allowed. Allowed: ${[...ALLOWED_EXTENSIONS].join(", ")}`,
      },
      400,
    );
  }
  if (file.size > MAX_MODEL_UPLOAD_BYTES) {
    return c.json({ detail: `File too large. Max size is ${MAX_MODEL_UPLOAD_BYTES} bytes.` }, 413);
  }

  const uniqueId = crypto.randomUUID().substring(0, 8);
  const safeStem = sanitizeUploadStem(file.name);
  const safeFilename = `${uniqueId}_${safeStem}${ext}`;
  const filePath = join(UPLOAD_DIR, safeFilename);

  try {
    await Bun.write(filePath, file);
    const urlPath = `/api/integrations/models/file/${safeFilename}`;
    return c.json({
      success: true,
      filename: safeFilename,
      url: urlPath,
      size: file.size,
    });
  } catch (err) {
    console.error("Failed to upload model:", err);
    return c.json({ detail: "Failed to upload model" }, 500);
  }
});

app.get("/models", async (c) => {
  try {
    const entries = await readdir(UPLOAD_DIR);
    const models = [];
    for (const name of entries) {
      const ext = extname(name).toLowerCase();
      if (!ALLOWED_EXTENSIONS.has(ext)) continue;
      const file = Bun.file(join(UPLOAD_DIR, name));
      models.push({
        filename: name,
        url: `/api/integrations/models/file/${name}`,
        size: file.size,
        type: ext === ".vrm" ? "vrm" : "image",
      });
    }
    return c.json({ models });
  } catch {
    return c.json({ models: [] });
  }
});

app.delete("/models/:filename", async (c) => {
  const filename = c.req.param("filename");
  if (filename.includes("..") || filename.includes("/") || filename.includes("\\")) {
    return c.json({ detail: "Invalid filename" }, 400);
  }

  const filePath = join(UPLOAD_DIR, filename);
  try {
    await unlink(filePath);
    return c.json({ success: true, message: `Deleted ${filename}` });
  } catch {
    return c.json({ detail: "File not found" }, 404);
  }
});

app.get("/models/file/:filename", async (c) => {
  const filename = c.req.param("filename");
  if (filename.includes("..") || filename.includes("/") || filename.includes("\\")) {
    return c.json({ detail: "Invalid filename" }, 400);
  }

  const filePath = join(UPLOAD_DIR, filename);
  const file = Bun.file(filePath);
  if (!(await file.exists())) {
    return c.json({ detail: "File not found" }, 404);
  }

  const ext = extname(filename).toLowerCase();
  const mediaTypes: Record<string, string> = {
    ".vrm": "model/gltf-binary",
    ".png": "image/png",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".gif": "image/gif",
    ".webp": "image/webp",
  };

  return new Response(file, {
    headers: { "Content-Type": mediaTypes[ext] ?? "application/octet-stream" },
  });
});

// ─── Animation Upload/List/Delete/Serve ───

app.post("/animations/upload", async (c) => {
  const formData = await c.req.formData();
  const file = formData.get("file") as File | null;
  if (!file || !file.name) {
    return c.json({ detail: "No file provided" }, 400);
  }

  const ext = extname(file.name).toLowerCase();
  if (!ALLOWED_ANIMATION_EXTENSIONS.has(ext)) {
    return c.json(
      {
        detail: `File type not allowed. Allowed: ${[...ALLOWED_ANIMATION_EXTENSIONS].join(", ")}`,
      },
      400,
    );
  }
  if (file.size > MAX_ANIMATION_UPLOAD_BYTES) {
    return c.json(
      {
        detail: `File too large. Max size is ${MAX_ANIMATION_UPLOAD_BYTES} bytes.`,
      },
      413,
    );
  }

  const uniqueId = crypto.randomUUID().substring(0, 8);
  const safeStem = sanitizeUploadStem(file.name);
  const safeFilename = `${uniqueId}_${safeStem}${ext}`;
  const filePath = join(ANIMATIONS_DIR, safeFilename);

  try {
    await Bun.write(filePath, file);
    return c.json({
      success: true,
      filename: safeFilename,
      url: `/api/integrations/animations/file/${safeFilename}`,
      size: file.size,
    });
  } catch (err) {
    console.error("Failed to upload animation:", err);
    return c.json({ detail: "Failed to upload animation" }, 500);
  }
});

app.get("/animations", async (c) => {
  try {
    const entries = await readdir(ANIMATIONS_DIR);
    const animations = [];
    for (const name of entries) {
      const ext = extname(name).toLowerCase();
      if (!ALLOWED_ANIMATION_EXTENSIONS.has(ext)) continue;
      const file = Bun.file(join(ANIMATIONS_DIR, name));
      animations.push({
        filename: name,
        url: `/api/integrations/animations/file/${name}`,
        size: file.size,
        type: ext.substring(1),
      });
    }
    return c.json({ animations });
  } catch {
    return c.json({ animations: [] });
  }
});

app.delete("/animations/:filename", async (c) => {
  const filename = c.req.param("filename");
  if (filename.includes("..") || filename.includes("/") || filename.includes("\\")) {
    return c.json({ detail: "Invalid filename" }, 400);
  }

  try {
    await unlink(join(ANIMATIONS_DIR, filename));
    return c.json({ success: true, message: `Deleted ${filename}` });
  } catch {
    return c.json({ detail: "File not found" }, 404);
  }
});

app.get("/animations/file/:filename", async (c) => {
  const filename = c.req.param("filename");
  if (filename.includes("..") || filename.includes("/") || filename.includes("\\")) {
    return c.json({ detail: "Invalid filename" }, 400);
  }

  const filePath = join(ANIMATIONS_DIR, filename);
  const file = Bun.file(filePath);
  if (!(await file.exists())) {
    return c.json({ detail: "File not found" }, 404);
  }

  const ext = extname(filename).toLowerCase();
  const mediaTypes: Record<string, string> = {
    ".fbx": "application/octet-stream",
    ".glb": "model/gltf-binary",
    ".gltf": "model/gltf+json",
  };

  return new Response(file, {
    headers: { "Content-Type": mediaTypes[ext] ?? "application/octet-stream" },
  });
});

export { app as integrationRoutes };
